flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/restore/ForStIncrementalRestoreOperation.java

```diff
@@ -0,0 +1,393 @@
+    private final CustomInitializationMetrics customInitializationMetrics;
+    private long lastCompletedCheckpointId;
+    private UUID backendUID;
+
+    private boolean isKeySerializerCompatibilityChecked;
+
+    public ForStIncrementalRestoreOperation(
+            String operatorIdentifier,
+            KeyGroupRange keyGroupRange,
+            CloseableRegistry cancelStreamRegistry,
+            ClassLoader userCodeClassLoader,
+            Map<String, ForStKvStateInfo> kvStateInformation,
+            StateSerializerProvider<K> keySerializerProvider,
+            Path forstBasePath,
+            File instanceRocksDBPath,
+            DBOptions dbOptions,
+            Function<String, ColumnFamilyOptions> columnFamilyOptionsFactory,
+            ForStNativeMetricOptions nativeMetricOptions,
+            MetricGroup metricGroup,
+            CustomInitializationMetrics customInitializationMetrics,
+            @Nonnull Collection<IncrementalRemoteKeyedStateHandle> restoreStateHandles) {
+        this.forstHandle =
+                new ForStHandle(
+                        kvStateInformation,
+                        instanceRocksDBPath,
+                        dbOptions,
+                        columnFamilyOptionsFactory,
+                        nativeMetricOptions,
+                        metricGroup);
+        this.operatorIdentifier = operatorIdentifier;
+        this.restoredSstFiles = new TreeMap<>();
+        this.lastCompletedCheckpointId = -1L;
+        this.backendUID = UUID.randomUUID();
+        this.customInitializationMetrics = customInitializationMetrics;
+        this.restoreStateHandles = restoreStateHandles;
+        this.cancelStreamRegistry = cancelStreamRegistry;
+        this.keyGroupRange = keyGroupRange;
+        this.forstBasePath = forstBasePath;
+        this.keySerializerProvider = keySerializerProvider;
+        this.userCodeClassLoader = userCodeClassLoader;
+    }
+
+    /**
+     * Root method that branches for different implementations of {@link
+     * IncrementalKeyedStateHandle}.
+     */
+    @Override
+    public ForStRestoreResult restore() throws Exception {
+
+        if (restoreStateHandles == null || restoreStateHandles.isEmpty()) {
+            return null;
+        }
+
+        logger.info(
+                "Starting ForSt incremental recovery in operator {}, target key-group range {}. State Handles={}",
+                operatorIdentifier,
+                keyGroupRange.prettyPrintInterval(),
+                restoreStateHandles);
+
+        List<IncrementalRemoteKeyedStateHandle> otherHandles =
+                new ArrayList<>(restoreStateHandles.size() - 1);
+        IncrementalRemoteKeyedStateHandle mainHandle =
+                chooseMainHandleAndCollectOthers(otherHandles);
+
+        StateHandleTransferSpec mainSpec =
+                new StateHandleTransferSpec(
+                        mainHandle, new Path(forstBasePath, ForStResourceContainer.DB_DIR_STRING));
```

Comment:
Why should the base one be downloaded into "db" folder?

Reply:
Thanks for review, fixed.